### PR TITLE
CorrDiff: Allreduced loss logging

### PR DESCRIPTION
**Pull request by mnabian**
# Modulus Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

**Review by yairchn — APPROVED** [review]
LGTM!

File: examples/generative/corrdiff/conf/config_train_base.yaml [hunk 1]
```diff
@@ -76,12 +76,8 @@ wandb_mode: online
   # Weights & biases mode [online, ofline, disabled]
 wandb_project:  "Modulus-Generative"
  # Weights & biases project
-wandb_group: "CorrDiff-DDP-Group"
- # Weights & biases group
-wandb_entity: "CorrDiff-DDP-Group"
+wandb_entity: "CorrDiff"
```
Comment (nbren12) [inline examples/generative/corrdiff/conf/config_train_base.yaml:79]: Is this actually the name of the "entity" for corrdiff? In the link below, "entity" is "nv-research-climate". Maybe we can just avoid setting an option here and let wandb do it's default behavior (entity=user name).

https://wandb.ai/nv-research-climate/edm-chaos/runs/jfw64cdl/overview?nw=nwusernbren12

File: examples/generative/corrdiff/training/training_loop.py [hunk 2]
```diff
@@ -276,15 +274,22 @@ def training_loop(
...
                 loss_accum += loss / num_accumulation_rounds
                 loss.backward()
-        wb.log({"training loss": loss_accum}, step=cur_nimg)
+
+        loss_sum = torch.tensor([loss_accum], device=device)
+        if dist.world_size > 1:
+            torch.distributed.all_reduce(loss_sum, op=torch.distributed.ReduceOp.SUM)
+        average_loss = loss_sum / dist.world_size
+        if dist.rank == 0:
+            wb.log({"training loss": average_loss}, step=cur_nimg)
```
Comment (nbren12) [inline examples/generative/corrdiff/training/training_loop.py:283]: Probably better to use the stats_collector instead to ensure compatibility with the stats.jsonl and avoid having two ways of logging stats. The stats_collector already handles the broadcast and averaging across samples.


In my edm code I added a `log_to_tensorboard` function to `training_stats.Collector`

```python
    def log_to_tensorboard(self, writer: SummaryWriter):
        info = self.as_dict()
        try:
            nimg = info["Progress/kimg"]["mean"] * 1000
        except KeyError:
            nimg = None

        for k, v in info.items():
            for moment in v:
                writer.add_scalar(f"{k}/{moment}", v[moment], global_step=nimg)
```

You can easily modify this to log with wandb instead. And then in training_loop.py


```
train_collector = Collector()
for batch in batch:
      train_collector.report("Loss/loss", loss)
      
train_collector.update()
train_collector.log_to_wandb()
```

And similarly for the test loss.

**Review by nbren12 — APPROVED** [review]
Fine for now, but would be cleaner to use the existing utilties, and I think will result in consistent values appearing in wandb and the stats.jsonl files.

File: examples/generative/corrdiff/training/training_loop.py (hunk 2 above)
Comment (nbren12) [reply examples/generative/corrdiff/training/training_loop.py:283]: see https://github.com/NVIDIA/modulus/blob/ac0eff476656a389028637fb7358c3704cfb2f47/examples/generative/corrdiff/training/training_loop.py#L402

Comment (nbren12) [reply examples/generative/corrdiff/training/training_loop.py:283]: you could event just put wandb logging at line 402

File: examples/generative/corrdiff/conf/config_train_base.yaml (hunk 1 above)
Comment (mnabian) [reply examples/generative/corrdiff/conf/config_train_base.yaml:79]: I think we will be creating a new wandb team, and can edit these configs based on that.

File: examples/generative/corrdiff/training/training_loop.py (hunk 2 above)
Comment (mnabian) [reply examples/generative/corrdiff/training/training_loop.py:283]: The problem is, averaging in stats collector doesn't work as expected, and it essentially logs the rank0 loss. We can revisit that in a separate PR.
